web/app/db.py

```diff
@@ -0,0 +1,233 @@
+    Helper function to allow for easy queries.
+    
+    :param query: The query to be executed.
+    :param args: The arguments to be passed into the query.
+    :param one: Whether the query should return all results or just the first. 
+    :return The results of the query.
+    """
+    cur = get_db().execute(query, args)
+    rv = cur.fetchall()
+    cur.close()
+    return (rv[0] if rv else None) if one else rv
+
+
+def close_db(e=None):
+    """
+    Closes the connection to the database. Called after every request.
+    """
+    db = g.pop('db', None)
+
+    if db is not None:
+        db.close()
+
+# Table Specific Functions
+def get_all_users():
+    """
+    Returns all users.
+
+    :return A dictionary of users with their ids as keys.
+    """
+    rows = query_db("SELECT * FROM user")
+
+    if rows:
+        return {row[0]: {"username": row[1], "preferences": row[2]} for row in rows}
+    else:
+        return {}
+
+def insert_user(username: str) -> Tuple[int, str]:
+    """
+    Inserts a new user. If the desired username is already taken,  
+    appends integers incrementally until an open name is found.
+
+    :param username: The desired username for the new user.
+    :return A tuple containing the id and name of the new user.
+    """
+    db = get_db()
+
+    new_user_id = None
+    count = 0
+    while new_user_id == None:
+        temp_name = username
+        
+        if count != 0:
+            temp_name += str(count)
+        try:
+            cur = db.execute('INSERT INTO user (username) VALUES (?)', [temp_name])
+            new_user_id = cur.lastrowid
+        except sqlite3.IntegrityError:
+            count += 1
+            continue
+    
+    db.commit()
+    cur.close()
+
+    return new_user_id, temp_name
+
+def get_ckpts(user_id: int):
+    """
+    Returns the checkpoints associated with the given user.
+    If no user_id is provided, return the checkpoints associated
+    with the default user.
+
+    :param user_id: The id of the user whose checkpoints are returned.
+    :return A list of checkpoints.
+    """
+    if not user_id:
+        user_id = 0
+
+    return query_db('SELECT * FROM ckpt WHERE associated_user = ' + str(user_id))
+
+def insert_ckpt(ckpt_name: str, 
+                associated_user: str, 
+                model_class: str, 
+                num_epochs: int) -> Tuple[int, str]:
+    """
+    Inserts a new checkpoint. If the desired name is already taken,  
+    appends integers incrementally until an open name is found.   
+
+    :param ckpt_name: The desired name for the new checkpoint.
+    :param associated_user: The user that should be associated with the new checkpoint.
+    :param model_class: The class of the new checkpoint.
+    :param num_epochs: The number of epochs the new checkpoint will run for.
+    :return A tuple containing the id and name of the new checkpoint.   
+    """
+    db = get_db()
+
+    new_ckpt_id = None
+    count = 0
+    while new_ckpt_id == None:
```

Comment:
I considered this, but I think since the query is different in each instance, it would be a bit of a pain to abstract out this functionality (query varies, name of variable that is changing from call to call varies, etc).